src/main/java/ti4/generator/UnitRenderGenerator.java

```diff
@@ -0,0 +1,545 @@
+    }
+
+    private int getCenteredDamageX(Point position, ImagePosition imagePosition, BufferedImage unitImage, BufferedImage dmgImage) {
+        return position != null
+            ? position.x + (unitImage.getWidth() / 2) - (dmgImage.getWidth() / 2)
+            : imagePosition.originalX() - (dmgImage.getWidth() / 2);
+    }
+
+    private int getCenteredDamageY(Point position, ImagePosition imagePosition, BufferedImage unitImage, BufferedImage dmgImage) {
+        return position != null
+            ? position.y + (unitImage.getHeight() / 2) - (dmgImage.getHeight() / 2)
+            : imagePosition.originalY() - (dmgImage.getHeight() / 2);
+    }
+
+    private boolean shouldSkipInvalidUnit(UnitKey unitKey) {
+        return unitKey == null || !Mapper.isValidColor(unitKey.getColor());
+    }
+
+    private boolean shouldHideJailUnit(Player frogPlayer, UnitKey unitKey) {
+        if (!ctx.isJail || frogPlayer == null) {
+            return false;
+        }
+        String colorID = Mapper.getColorID(frogPlayer.getColor());
+        return !ctx.showJail && !unitKey.getColorID().equals(colorID);
+    }
+
+    private String getUnitPath(UnitKey unitKey) {
+        String unitPath = Tile.getUnitPath(unitKey);
+        if (unitPath == null) return null;
+
+        // Handle bulk unit replacements
+        unitPath = switch (unitKey.getUnitType()) {
+            case Fighter -> unitPath.replace(Constants.COLOR_FF, Constants.BULK_FF);
+            case Infantry -> unitPath.replace(Constants.COLOR_GF, Constants.BULK_GF);
+            default -> unitPath;
+        };
+
+        Player player = game.getPlayerByColorID(unitKey.getColorID()).orElse(null);
+        if (player == null) return unitPath;
+
+        // Handle special unit cases
+        if (unitKey.getUnitType() == UnitType.Spacedock && player.ownsUnitSubstring("cabal_spacedock")) {
+            return unitPath.replace("sd", "csd");
+        }
+
+        // Handle special unit replacements
+        return switch (unitKey.getUnitType()) {
+            case Lady -> unitPath.replace("lady", "fs");
+            case Cavalry -> {
+                boolean hasM2Tech = game.getPNOwner("cavalry") != null &&
+                    game.getPNOwner("cavalry").hasTech("m2");
+                String version = hasM2Tech ? "Memoria_2.png" : "Memoria_1.png";
+                yield resourceHelper.getUnitFile(version);
+            }
+            default -> unitPath;
+        };
+    }
+
+    private BufferedImage getUnitDecal(Player player, UnitKey unitKey) {
+        if (player == null) return null;
+        return ImageHelper.read(resourceHelper.getDecalFile(player.getDecalFile(unitKey.asyncID())));
+    }
+
+    private BufferedImage getSpoopyImage(UnitKey unitKey, Player player) {
+        BufferedImage spoopy = null;
+
+        // Random spoopy warsun
+        if (unitKey.getUnitType() == UnitType.Warsun && ThreadLocalRandom.current().nextInt(1000) == 0) {
```

Comment:
just because I see it, use CalendarHelper.isNearHalloween() to control whether spooky warsun chance is 1/10 or 1/1000